Add NUCLEO-L4R5ZI Board

psidhu opened the pull request:
Adds support for aforementioned board. Based on the STM32L4 family.

Tested blinky and hello world so far.

psidhu committed:
arm: st: add stm32l4r5xx support

New parts from ST. See http://www.st.com/en/microcontrollers/stm32l4r5-s5.html
for more details.

Signed-off-by: Pushpal Sidhu <psidhu.devel@gmail.com>

psidhu committed:
arch: arm: mpu: add support for 640k SRAM

The STM32L4+ parts have 640k of SRAM.

Signed-off-by: Pushpal Sidhu <psidhu.devel@gmail.com>

psidhu committed:
boards: arm: Add nucleo_l4r5zi board support

This board features a STM32L4R5ZI processor.

Signed-off-by: Pushpal Sidhu <psidhu.devel@gmail.com>

dbkinder commented on arch/arm/soc/st_stm32/stm32l4/Kconfig.defconfig.stm32l4r5xx:
```diff
@@ -0,0 +1,18 @@
+# Kconfig - ST Microelectronics STM32L4R5xx MCU
+#
+# Copyright (c) 2017 Pushpal Sidhu
```
should all these headers be updated to say 2018 now?

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+
+- STM32 microcontroller in LQFP144 package
+- Two types of extension resources:
+  - Arduino Uno V3 connectivity
```
add a blank line before a sublist like this

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+  - ST morpho extension pin headers for full access to all STM32 I/Os
+- On-board ST-LINK/V2-1 debugger/programmer with SWD connector
+- Flexible board power supply:
+  - USB VBUS or external source(3.3V, 5V, 7 - 12V)
```
blank line here too

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+
+The STM32L4R5ZI SoC provides the following hardware IPs:
+
+- Ultra-low-power with FlexPowerControl (down to 130 nA Standby mode and 100 uA/MHz run mode)
```
keep line length < 80 chars please (unless unavoidable)

Not sure what "100 uA/MHz" means, if it's "100 uA and 100 MHz" it would be better to write it out as such (without a / )

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+The STM32L4R5ZI SoC provides the following hardware IPs:
+
+- Ultra-low-power with FlexPowerControl (down to 130 nA Standby mode and 100 uA/MHz run mode)
+- Core: ARM |reg| 32-bit Cortex |reg|-M4 CPU with FPU, adaptive real-time accelerator (ART Accelerator™) allowing 0-wait-state execution from Flash memory, frequency up to 120 MHz, MPU, 150 DMIPS/1.25 DMIPS/MHz (Dhrystone 2.1), and DSP instructions
```
OK to leave off the TM mark here, and keep line length < 80 chars

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+- Ultra-low-power with FlexPowerControl (down to 130 nA Standby mode and 100 uA/MHz run mode)
+- Core: ARM |reg| 32-bit Cortex |reg|-M4 CPU with FPU, adaptive real-time accelerator (ART Accelerator™) allowing 0-wait-state execution from Flash memory, frequency up to 120 MHz, MPU, 150 DMIPS/1.25 DMIPS/MHz (Dhrystone 2.1), and DSP instructions
+- Clock Sources:
+  - 4 to 48 MHz crystal oscillator
```
add a blank line before this sublist

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+- RTC with HW calendar, alarms and calibration
+- Up to 24 capacitive sensing channels: support touchkey, linear and rotary touch sensors
+- Advanced graphics features
+  - Chrom-ART Accelerator™ (DMA2D) for enhanced graphic content creation
```
blank line here (and drop the TM)

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+System Clock
+------------
+
+Nucleo L4R5ZI System Clock could be driven by internal or external oscillator,
```
``by an internal``

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+------------
+
+Nucleo L4R5ZI System Clock could be driven by internal or external oscillator,
+as well as main PLL clock. By default System clock is driven by PLL clock at 120MHz,
```
``as the main``   and   ``the System clock``  and  ``the PLL clock``

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+
+Nucleo L4R5ZI System Clock could be driven by internal or external oscillator,
+as well as main PLL clock. By default System clock is driven by PLL clock at 120MHz,
+driven by 16MHz high speed internal oscillator.
```
``by a 16MHz``

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+
+Nucleo L4R5ZI board includes an ST-LINK/V2-1 embedded debug tool interface.
+
+You currently need bleeding edge OpenOCD plus a few patches (build your own OpenOCD):
```
bleeding edge?  better would be to say which version (or greater) is needed.  Are the two bullet points below sufficient instructions to build a working OpenOCD?  If not, please provide instructions.

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+ * http://openocd.zylin.com/#/c/4310/
+ * http://openocd.zylin.com/#/c/4313/
+
+Either replace the OpenOCD binary in the Zephyr SDK, or execute it on
```
not sure what "execute it on your own" means?  Since your examples below assume OpeOCD has been replaced in the SDK, it would be best to explain how to do this.

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+ * http://openocd.zylin.com/#/c/4313/
+
+Either replace the OpenOCD binary in the Zephyr SDK, or execute it on
+your own. Expect this in a future version of the Zephyr SDK.
```
Is there a reported Issue against the SDK that captures this work?

dbkinder commented on boards/arm/nucleo_l4r5zi/dts.fixup:
```diff
@@ -0,0 +1,37 @@
+/* This file is a temporary workaround for mapping of the generated information
+ * to the current driver definitions.  This will be removed when the drivers
```
What's going to remind someone to remove this workaround?  Is there an issue reported to track this?

erwango replied:
.fixup files are documented on http://docs.zephyrproject.org/devices/dts/device_tree.html
This is also visible at project level, but I haven't seen any GH issue to track this activity

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+  - MIPI® DSI Host controller with two DSI lanes running at up to 500 Mbits/s each
+  - LCD-TFT controller
+- 16x timers
+  - 2 x 16-bit advanced motor-control
```
blank line needed here (I'll stop mentioning this...)

dbkinder reviewed (changes requested):
suggested edits

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+  - 2x low-power 16-bit timers (available in Stop mode)
+  - 2x watchdogs
+  - SysTick timer
+- Up to 136 fast I/Os, most 5 V-tolerant, up to 14 I/Os with independent supply down to 1.08 V 
```
shippable reports a space at the end of this line that should be deleted

dbkinder commented on boards/arm/nucleo_l4r5zi/doc/nucleol4r5zi.rst:
```diff
@@ -0,0 +1,252 @@
+- True random number generator
+- CRC calculation unit, 96-bit unique ID
+- 8- to 14-bit camera interface up to 32 MHz (black and white) or 10 MHz (color)
+- Development support: serial wire debug (SWD), JTAG, Embedded Trace Macrocell (ETM) 
```
and an extra space at the end of this line too

galak commented:
Can you pull out 'drivers: serial: uart_stm32: fix call to LL_USART_SetBaudRate' into a separate PR.

erwango commented on boards/arm/nucleo_l4r5zi/dts.fixup:
```diff
@@ -0,0 +1,37 @@
+#define CONFIG_UART_STM32_PORT_5_BAUD_RATE	ST_STM32_USART_40005000_CURRENT_SPEED
+#define CONFIG_UART_STM32_PORT_5_IRQ_PRI	ST_STM32_USART_40005000_IRQ_0_PRIORITY
+#define CONFIG_UART_STM32_PORT_5_NAME		ST_STM32_USART_40005000_LABEL
+#define PORT_5_IRQ				ST_STM32_USART_40005000_IRQ_0
```
Can you add I2C as well?

psidhu replied:
Sure

erwango commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.yaml:
```diff
@@ -0,0 +1,7 @@
+identifier: nucleo_lr5zi
```
nucleo_l4r5zi

erwango commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.yaml:
```diff
@@ -0,0 +1,7 @@
+identifier: nucleo_lr5zi
+name: NUCLEO-LR5ZI
```
NUCLEO-L4R5ZI

erwango commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi_defconfig:
```diff
@@ -0,0 +1,64 @@
+CONFIG_ARM=y
```
Could you align this file on recent #5515 ?

erwango commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.dts:
```diff
@@ -0,0 +1,33 @@
+	};
+};
+
+&usart2 {
```
This would be usart1.
Then do you want to activate USART 1?
This is not enabled in _defconfig file

psidhu replied:
Oh shoot, I might have missed that, thanks.

psidhu commented:
@dbkinder Thanks for reviewing that. First time doing anything with rst files.

galak commented:
Any updates on this?

psidhu commented:
Not yet, I haven't been able to work on anything zephyr related for awhile. I did, however, pull out one commit into a separate PR (#5814) as requested.

I'm hoping to get back to this in the coming weeks.

galak commented:
@psidhu just checking in on this

galak commented:
@psidhu any update?  If not going to close this for now

psidhu commented:
@galak Yeah, I have this up to date, but since there's no actual OpenOCD support, I've been holding off pushing it. Would it be OK to get this into zephy, but explain that 'make flash' won't work?

I can give an alternative like using the STM flash program (GUI program).

galak commented:
> @galak Yeah, I have this up to date, but since there's no actual OpenOCD support, I've been holding off pushing it. Would it be OK to get this into zephy, but explain that 'make flash' won't work?
> 
> I can give an alternative like using the STM flash program (GUI program).

OCD/flashing support isn't required for acceptance.

psidhu commented:
Gotcha. I'll clean it up a bit then push.

psidhu commented:
recheck

erwango commented:
Hi @psidhu ,

Can you rebase vs latest device tree changes?
Cf other stm32 boards.

Thanks

psidhu commented:
@erwango done

erwango reviewed (changes requested):
@psidhu : Last minor comment otherwise looks fine to me

psidhu commented:
@erwango Hi, I'm a little unsure of what you're asking for. The only comment I see is:
> Could you align this file on recent #5515 ?

Did I perhaps miss something? I thought I had done this.

psidhu commented:
@erwango ok, I think I see what you meant. I fixed the stm32l4r5.dtsi file pin-controller.

erwango reviewed (changes requested):
Very last comments and we're good.
"boost mode" would require more detailed documentation.

psidhu commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.yaml:
```diff
@@ -0,0 +1,11 @@
+  - zephyr
+  - gccarmemb
+supported:
+  - pwm
```
done

psidhu commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.dts:
```diff
@@ -0,0 +1,72 @@
+	pinctrl-names = "default";
+	status = "ok";
+};
+
```
done

psidhu commented on boards/arm/nucleo_l4r5zi/Kconfig.defconfig:
```diff
@@ -0,0 +1,37 @@
+
+endif # UART_CONSOLE
+
+if SPI
```
done

psidhu commented:
@erwango recheck. I also added USB since it was merged.

erwango commented:
> @erwango recheck. I also added USB since it was merged.

Then you'll need to document it :).
Also, please add USB_DEVICE_NETWORK_ECM as default network interface (cf nucleo_f412zg).

psidhu commented:
@erwango Sorry for the long break, I went on hiatus. I've pushed the requested changes except I used EEM instead of ECM.

erwango commented:
recheck

galak commented:
Any plans to update this?

erwango commented:
@psidhu , can you rebase and get CI pass?

galak commented:
> @psidhu , can you rebase and get CI pass?

ping

psidhu committed:
flash: stm32l4x: Correct getting page number from STM32L4R/S devices

These devices increase the page size from 2k (standard STM32L4) to 4K,
so we must account for this.

Signed-off-by: Pushpal Sidhu <psidhu.devel@gmail.com>

psidhu commented:
@erwango I've rebased. CI is only failing on some SPI/Logging thing which I haven't touched (unless I need to enable something): https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/23647/3/tests

superna9999 commented:
OK for the flash patch

erwango commented:
recheck

erwango commented on boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.dts:
```diff
@@ -0,0 +1,106 @@
+		sw0 = &user_button;
+	};
+};
+
```
Missing arduino connectors here:
```
arduino_i2c: &i2cx {};
arduino_spi: &spiy {};
arduino_serial: &usartz {};
```

psidhu replied:
@erwango Added. Please recheck.

erwango reviewed (approved):
Approved on my side.
Let's fix comments from @dbkinder and get this merged.
I've raised #10380 for CI issue.

psidhu commented:
@dbkinder Can you please review the documentation again?

dbkinder reviewed (approved):
+1 for doc changes, thanks
